scripts/cleanup_analysis_entries.py

```diff
@@ -0,0 +1,259 @@
+
+    return (analyses, ids_to_delete, duplicate_analyses_ids)
+
+def check_all_analyses_belongs_to_project(analyses: list[dict], project_id_map: dict):
+    # Check the analyses ['project'] matches the bucket name, `cpg-{dataset}-{main/test}`
+
+    # bad_mappings = [] # Use logger for this
+
+    for analysis in analyses:
+        project_id = analysis['project']
+        project_name = project_id_map[project_id]
+        dataset_name = get_path_components_from_path(analysis['output'])['dataset']
+        if dataset_name == project_name:
+            continue
+        else:
+            pass
+            # bad_mappings.append(analysis["id"])
+            # print(f'Analysis ID: {str(analysis["id"])} has project ID: {project_id} which maps to: {project_name}, but analysis dataset is: {dataset_name}')
+            # raise warnings.warn(f'Analysis ID: {str(analysis["id"])} has project ID: {project_id} which maps to: {project_name}, but analysis dataset is: {dataset_name}')
+
+
+def validate_paths(paths: iter) -> dict[str, bool]:
+    # Check if a path to an analysis object exists in the bucket
+    # return a dictionary of valid paths 
+
+    paths_by_bucket = defaultdict(list)
+    # iterate through all subdirectories/paths in a bucket
+    for p in paths:
+        bucket_name = p.removeprefix('gs://').split('/', maxsplit=1)[0]
+        paths_by_bucket[bucket_name].append(p)
+
+
+    valid_paths = {}
+    invalid_paths = {}
+    # Check if each subdirectory/path in a bucket is valid or not
+    for bucket_name, paths in paths_by_bucket.items():
+        vp, ivp = validate_paths_for_bucket(bucket_name, paths)
+        valid_paths.update(vp)
+        invalid_paths.update(ivp)
+
+    path_exists = {'valid': valid_paths, 'invalid': invalid_paths}
+    
+    return path_exists
+
+def validate_paths_for_bucket(bucket_name, paths: list[str]) -> tuple[set[str], set[str]]:
+    # group by path prefix
+
+    path_set = {}
+    for p in paths:
+        # subdir = p.removeprefix(f'gs://{bucket_name}/')
+        pc = get_path_components_from_path(p)
+        path_set.add(pc['subdir'])
+
+    files_by_subdir = {}
+    for subdir in path_set:
+        # don't ever read a directory with a path component that ends in [.ht, .mt]
+        if '.mt' in subdir or '.ht' in subdir:
+            continue
+        
+        files_by_subdir[subdir] = get_paths_for_subdir(client, bucket_name, subdir) 
+
+    # paths = all paths found in Metamist for a given bucket name from analysis['output']
+    # files_by_subdir = all paths found in gcloud storage for a given bucket grouped by bucket subdir
+    files_set = set([e for sublist in files_by_subdir.values for e in sublist])
+
+    # Has analysis entry but file moved or deleted:
+    analysis_entries_missing_files = path_set.difference(files_set)
+
+    # valid entries are all analysis output paths that intersect with all gcp output paths
+    valid_entries = path_set.intersect(files_set)
```

Comment:
```suggestion
    valid_entries = path_set.intersection(files_set)
```

I might be wrong about this...